### ISSUE TITLE
Accumulating TCP-Router HAProxy <defunct> instances

### ISSUE THREAD

**Issue by x6j8x** [cloudfoundry/routing-release#125]
## Issue

Since we've updated from `cf-deployment` 3.6 to 4.5 - which included the bpm-ization of tcp-router as far as I can tell - we're having troubles with failing `tcp-router` nodes.

The typical picture is that we're seeing are thousands of `defunct` haproxy processes on the failing nodes.

```
init-+-auditd-+-audispd---{audispd}
     |        `-{auditd}
     |-blackbox---7*[{blackbox}]
     |-bosh-dns---8*[{bosh-dns}]
     |-bosh-dns-namese---7*[{bosh-dns-namese}]
     |-cf-tcp-router-+-4784*[haproxy]
     |               `-9*[{cf-tcp-router}]
```

From what I gather from Google wisdom ("haproxy defunct container") is that Haproxy behaves differently in a container then a "plain" linux system when reloading the configuration. The old processes are not cleaned up and stay around.

## Context

We're using tcp-router in production and every day some instances fail in any of our 6 geographic regions. 

## Possible Fix

https://github.com/QubitProducts/bamboo/issues/31#issuecomment-57165606

**Comment by aaronshurley** [comment]
Thank you for creating this issue. We'll have to investigate this further to provide a proper fix.

/cc @shalako @shubhaat for prioritization

**Comment by x6j8x** [comment]
@rosenhouse 

Might be related or maybe not but IIRC `route-registrar` was also converted to BPM recently. 

We're seeing this on our `api` (cloud_controller) instances: 

```
init-+-auditd-+-audispd---{audispd}
     |        `-{auditd}
     |-route-registrar-+-661*[bash]
     |                 `-10*[{route-registrar}]
```

**Comment by aaronshurley** [comment]
@x6j8x I was able to see the HAProxy defunct processes on the `tcp-router` instances but I don't see the same for `route-registrar` on the `api` instances. This might be worth tracking in a different github issue. Could you share what the output you see for one of the `route-registrar` lines are, is it the same `defunct` issue? Could you share a bit more about your deployment so that we can try to replicate it?

**Comment by x6j8x** [comment]
@aaronshurley we have a pretty plain `cf-deployment 4.5` environment. 

We use `tcp-router` for some internal functionalities hosted on CF with no more then 5-6 routes.

Regarding the `route-registrar` issue. We see `bash <defunct>` entries in the process list. As we had some issues with the api route-registrar health check script we've disabled it for now. So let's put this aside for now. I'll open a new issue once we're seeing it again.

**Comment by x6j8x** [comment]
@aaronshurley maybe the SIGCHLD handling should be applied everywhere where external processes are started from a executable running in a bpm container (like route-registrar + healthcheck scripts).

**Comment by aaronshurley** [comment]
@x6j8x that's a great point. We created a (story)[https://www.pivotaltracker.com/story/show/161569487] to make this fix in route-registrar. If you find any other occurrences, please open up github issues in the appropriate repos. Thanks!

Since this original issue has been accepted, it should be cut with our next final release. We're going to close this issue.

**Comment by x6j8x** [comment]
@aaronshurley Can you give any indication when the next release will be cut? It's really tedious to regulary have to restart all those tcp-router instances across our 6 regions.

**Comment by shubhaat** [comment]
@x6j8x We are so sorry it has taken us so long to cut a release. We are still blocked by performance reports, but I will discuss this with the team today and see how we can prioritize cutting a release.

**Comment by shubhaat** [comment]
Hi @x6j8x - We cut a release this morning (routing release [0.183.0](https://github.com/cloudfoundry/routing-release/releases/tag/0.183.0))
